terminal/command.go

```diff
@@ -202,18 +208,124 @@ func thread(client service.Client, args ...string) error {
 }
 
 func goroutines(client service.Client, args ...string) error {
+	state, err := client.GetState()
+	if err != nil {
+		return err
+	}
 	gs, err := client.ListGoroutines()
 	if err != nil {
 		return err
 	}
 	fmt.Printf("[%d goroutines]\n", len(gs))
 	for _, g := range gs {
-		fmt.Printf("Goroutine %s\n", formatGoroutine(g))
+		prefix := "  "
+		if g.ID == state.SelectedGoroutine.ID {
+			prefix = "* "
+		}
+		fmt.Printf("%sGoroutine %s\n", prefix, formatGoroutine(g))
+	}
+	return nil
+}
+
+func goroutine(client service.Client, args ...string) error {
+	switch len(args) {
+	case 0:
+		return printscope(client)
+
+	case 1:
+		gid, err := strconv.Atoi(args[0])
+		if err != nil {
+			return err
+		}
+
+		oldState, err := client.GetState()
+		if err != nil {
+			return err
+		}
+		newState, err := client.SwitchGoroutine(gid)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("Switched from %d to %d (thread %d)\n", oldState.SelectedGoroutine.ID, gid, newState.CurrentThread.ID)
+		return nil
+
+	default:
+		return scopePrefix(client, "goroutine", args...)
+	}
+}
+
+func frame(client service.Client, args ...string) error {
+	return scopePrefix(client, "frame", args...)
+}
+
+func scopePrefix(client service.Client, cmdname string, pargs ...string) error {
+	fullargs := make([]string, 0, len(pargs)+1)
+	fullargs = append(fullargs, cmdname)
+	fullargs = append(fullargs, pargs...)
+
+	scope := api.EvalScope{-1, 0}
+	lastcmd := ""
+
+	callFilterSortAndOutput := func(fn scopedFilteringFunc, fnargs []string) error {
+		outfn := filterSortAndOutput(func(client service.Client, filter string) ([]string, error) {
+			return fn(client, scope, filter)
+		})
+		return outfn(client, fnargs...)
 	}
+
+	for i := 0; i < len(fullargs); i++ {
+		lastcmd = fullargs[i]
+		switch fullargs[i] {
+		case "goroutine":
+			n, err := strconv.Atoi(fullargs[i+1])
+			if err != nil {
+				return fmt.Errorf("Invalid argument to goroutine, expected integer")
+			}
+			scope.GoroutineID = int(n)
+			i++
+		case "frame":
+			n, err := strconv.Atoi(fullargs[i+1])
```

Comment:
It's this line that is panic'ing with an "index out of range" error, when you invoke `frame` without any arguments.